flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStore.java

```diff
@@ -166,20 +190,38 @@ public TaskStateSnapshot retrieveLocalState(long checkpointID) {
 	 * Disposes the state of all local snapshots managed by this object.
 	 */
 	public void dispose() {
+
+		Collection<Map.Entry<Long, TaskStateSnapshot>> statesCopy;
+
 		synchronized (lock) {
-			for (Map.Entry<Long, TaskStateSnapshot> entry : storedTaskStateByCheckpointID.entrySet()) {
-				discardStateObject(entry.getValue(), entry.getKey());
-			}
 			discarded = true;
+			statesCopy = new ArrayList<>(storedTaskStateByCheckpointID.entrySet());
 		}
+
+		discardExecutor.execute(() -> {
```

Comment:
I think it would be good for these kind of asynchronous clean up operations to return a `CompletableFuture`. This future could then be returned by the `dispose` method. The benefit would be that the caller would know when the clean up has completed and, thus, would be safe to shut down the `discardExecutor`.

Reply:
@tillrohrmann `ExecutorService` has a shutdown method (the one we call) that waits until all submitted futures are completed. Do you see any additional value in tracking the future completion individually?

Reply:
👍